modules/partition-replicator/src/main/java/org/apache/ignite/internal/partition/replicator/PartitionReplicaLifecycleManager.java

```diff
@@ -1240,4 +1278,52 @@ private void cleanUpPartitionsResources(Set<ReplicationGroupId> partitionIds) {
             LOG.error("Unable to clean zones resources", e);
         }
     }
+
+    /**
+     * Lock the zones replica list for any changes. {@link #hasLocalPartition(ZonePartitionId)} must be executed under this lock always.
+     *
+     * @param zoneId Zone id.
+     * @return Stamp, which must be used for further unlock.
+     */
+    public long lockZoneIdForRead(int zoneId) {
+        AtomicLong stamp = new AtomicLong();
+
+        zonePartitionsLocks.compute(zoneId, (id, l) -> {
+            if (l == null) {
+                l = new StampedLock();
+            }
+
+            stamp.set(l.readLock());
+
+            return l;
+        });
+
+        return stamp.get();
+    }
+
+    /**
+     * Unlock zones replica list.
+     *
+     * @param zoneId Zone id.
+     * @param stamp Stamp, produced by the according {@link #hasLocalPartition(ZonePartitionId) call.}
+     */
+    public void unlockZoneIdForRead(int zoneId, long stamp) {
```

Comment:
I believe that it worth add simple unit tests for lock/unlock.

Reply:
I thought about, but to be honest - I couldn't imagine the test, which is not testing the logic of StampedLock as a result, it will be strange. But at the same time, possible complex logic races and other issues will be tested by our usual tests, when we switch main to collocation path.